### ISSUE TITLE
Data Export: Exported columns depend on first data point, and problems with missing columns

### ISSUE THREAD

**Issue by CipSoft-Components** [amcharts/amcharts4#681]
When exporting the charts data to XLSX or CSV, only columns found in the first data point are exported.
e.g data:
```
[
  {"date": "2018-12-01"},
  {"date": "2018-12-02", "val": "1"}
]
```
exported CSV:
```
date
2018-12-01
2018-12-02
```

The "val" column is missing.
For the same example you can force exporting known columns by setting `chart.exporting.dataFields = {val: 'Value', date: 'Date'}`, but then for the missing value data point, the date is in the wrong column if the "val" column is the first one (in CSV the comma is missing).
Exported CSV:
```
Value,Date
2018-12-01
1,2018-12-02
```

**Comment by martynasma** [comment]
You're right. The export looks at the first data point to determine the columns.

Using `dataFields` should be the solution. However, there's probably a bug, as you described, with order of columns being messed up.

We'll let you know when it's fixed.

**Comment by martynasma** [comment]
This issue was fixed in dev version, and will be released soon.

We'll notify you here when that happens.

In addition to `dataFields` being respected when exporting to CSV/Excel, it now also will have an option `emptyAs` which will allow setting what to replace empty values with.

**Comment by CipSoft-Components** [comment]
Thank you! The `emptyAs` is also a good addition.
Is it also possible to add an option, which automatically export all dataFields, at least the ones shown in in the chart? Until now I didn't tried out the legend of amCharts 4, but this should also know all dataFields from the series, or do I also have to set them manually?

**Comment by martynasma** [comment]
Hm, exporting all values would require us to iterate through all of the data.

Generating data fields based on Series might be an option, but I'm kinda reluctant. Export operation is universal (it does not necessarily have to be a chart), whereas Series/Legend is tied into specific chart types, etc.

I'll give it some more thought, tho. Thanks for the suggestions!

**Comment by martynasma** [comment]
OK, implemented field search in all of the data. Let's see where that gets us :)

**Comment by martynasma** [comment]
Fixed in [4.0.9](https://github.com/amcharts/amcharts4/releases/tag/4.0.9).

## [4.0.9] - 2018-12-19

### Added
- `snapToSeries` added to `XYCursor`. Allows setting series to which cursor lines should be snapped. Works if one of the axis of the series is `DateAxis` or `CategoryAxis`. Won't work if both axes are `ValueAxis`.
- New example: Pie charts in Columns using JSON config.
- Export: `emptyAs` added to both CSV and Excel export options. If set, missing values will be replaced with this value (default is empty string).
- New event `"beforedisposed"` added to `Sprite`. Kicks in right before the element starts to dismantle itself.
- Italian translation.

### Changed
- Export: if no `dataFields` specify it will now look for fields in all of the data, not just the first line.
- When creating a new state on an object, it will not automatically propagate itself on object's clones if `applyOnClones = true`.
- Setting `Container`'s `setStateOnChildren` will now propage to the new value to its existing clones if `applyOnClones = true`.
- In case `deepInvalidate()` is called on `Container`, it forces labels to redraw.
- `filters` on `SpriteState` is now `List` (was `ListTemplate`). It's better for performance which does not require creating Filter template object.

### Fixed
- Clicking a hoverable/clickable object was resultin in "over" and "out" events generated next to "hit".
- JSON: properties to existing `List` items were not being applied properly.
- Using geodata in Ember app was causing errors (Issue 672).
- Export to CSV/Excel was following first line of data values. It now respects `dataFields` setting, and will replace missing values with `emptyAs` (new options setting).
- Wheel-scrolling in MacOS Safari was very slow.
- HTML-based tooltips were too small to fit the actual contents on MaCOS Safari.
- `DateFormatter` was not parsing `"i"` format as local timezone, rather than UTC.
- Sometimes bullets with `Rectangle` element in them were not positioned properly.
- If category was named `"undefined"`, it was messing up label positions on `CategoryAxis`.
- State's property value was not being applied if current value of the property was `undefined`.
- Performance of `PieChart` and `RadarChart` improved dramatically under IEs.
- In some cases `"ready"` event was not being fired.
- `FunnelSeries` with 0 (zero) value slices was not rendered correctly.

[Full change log](https://github.com/amcharts/amcharts4/blob/master/dist/es2015/CHANGELOG.md).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
